lib/elixir/lib/macro.ex

```diff
@@ -98,6 +98,103 @@ defmodule Macro do
     end
   end
 
+  # Classifies the given atom into one of five categories:
+  #
+  #   * :boolean_or_nil - the atoms true, false, nil
+  #
+  #   * :alias - a valid Elixir alias, like Foo, Foo.Bar and so on
+  #
+  #   * :callable - an atom that can be used as a function call after the
+  #     . operator (for example, :<> is callable because Foo.<>(1, 2, 3) is valid
+  #     syntax); this category includes identifiers like :foo
+  #
+  #   * :non_callable - an atom that cannot be used as a function call after the
+  #     . operator (for example, :<<>> is not callable because Foo.<<>> is a
+  #     syntax error); this category includes atoms like :Foo, since they are
+  #     valid identifiers but they need quotes to be used in function calls
+  #     (Foo."Bar")
+  #
+  #   * :atom - any other atom (these are usually escaped when inspected, like
+  #     :"foo and bar")
+  @doc false
+  def classify_atom(atom_or_string)
+
+  unary_ops_as_strings = :lists.map(&:erlang.atom_to_binary(&1, :utf8), @unary_ops)
+  binary_ops_as_strings = :lists.map(&:erlang.atom_to_binary(&1, :utf8), @binary_ops)
+
+  def classify_atom(atom) when is_atom(atom) do
+    classify_atom(Atom.to_string(atom))
+  end
+
+  def classify_atom(string) when is_binary(string) do
+    cond do
+      string in ["nil", "true", "false"] ->
+        :boolean_or_nil
```

Comment:
As @lexmag rightfully pointed out, this could have been split in two, `:boolean` and `:nil`, not have this `:boolean_or_nil` result. I started out with that initially actually, but then changed to `:boolean_or_nil`. The rationale for the change was: this `classify_atom/1` function is not public (so we are in control of it and don't need to be unnecessarily flexible to accommodate users needs) and for now it is only used for inspecting one way or another. When inspecting, `nil`, `true`, and `false` all belong to the same category of special atoms, that is, atoms that are not aliases but still inspected without the leading `:`.

After this whole discussion (which showed this is not the clearest design) I came up with what I think is an even better idea: we can special case `nil` and booleans only in `Inspect.Atom.inspect/1`, because they are not special when considered as function names (that is, they are just `:callable`. I will do this in a separate commit.